landscape/client/manager/snapmanager.py

```diff
@@ -0,0 +1,226 @@
+                result = e.json["result"]
+                logging.error("Error installing snap '%s': %s", name, result)
+                errors[(name, revision, channel)] = result
+
+        return self._check_statuses(installing, opid, errors)
+
+    def _handle_remove_snaps(self, message):
+        """
+        Removes the snaps indicated in `message` using the snap interface.
+        """
+        snaps = message["snaps"]
+        opid = message["operation-id"]
+        snap_http = SnapHttp()
+        errors = {}
+        removing = []
+
+        logging.info("Removing snaps: %s", snaps)
+
+        # See comment in `_handle_install_snaps` for reasoning behind this sync
+        # approach.
+        for snap in snaps:
+            name = snap["name"]
+
+            try:
+                response = snap_http.remove_snap(name)
+
+                if "change" not in response:
+                    logging.error(
+                        "Error removing snap '%s': %s",
+                        name,
+                        response["result"]["message"],
+                    )
+                    errors[name] = str(response)
+
+                removing.append((response["change"], name))
+            except SnapdHttpException as e:
+                result = e.json["result"]
+                logging.error("Error removing snap '%s':%s", name, result)
+                errors[name] = result
+
+        return self._check_statuses(removing, opid, errors)
+
+    def _check_statuses(self, change_ids, opid, errors):
+        """
+        Repeatedly polls for the status of each change in `change_ids`
+        until all are done.
+
+        Because there might be multiple changes kicked off at once, we use
+        `gatherResults` to wait until all of them are completed (successfully
+        or otherwise).
+        """
+        deferred_results = defer.gatherResults(
+            [self._get_status(cid, name) for cid, name in change_ids],
+            consumeErrors=True,
+        )
+
+        deferred_results.addCallback(
+            lambda results: self._respond(results, opid, errors),
+        )
+
+        return deferred_results
+
+    def _get_status(self, change_id, snap_name):
+        """
+        Uses the snap interface to retrieve the status of a snap change.
+        Polls every `interval` seconds a maximum of `attempts` times, as
+        configured in the registry.
+        """
+        counter = 0
+        attempts = getattr(self.registry.config, "snapd_poll_attempts", 5)
+        interval = getattr(self.registry.config, "snapd_poll_interval", 15)
+        snap_http = SnapHttp()
+
+        def get_status():
+            """
+            Looping function that stashes results in loop.result. Exits when a
+            non-incomplete snap change status is found, or upon an error.
+            """
+            nonlocal counter
+            counter += 1
+
+            logging.debug("Polling snapd for status of change %s", change_id)
```

Comment:
Some of these could be in, info(), would make it easier to go through broker.log since, debug is not enabled by default